cmd/test-server/main.go

```diff
@@ -143,41 +128,271 @@ func echoHandler(w http.ResponseWriter, r *http.Request) {
-	data := stats.LoadRawData(arr)
-	min, _ := stats.Min(data)
-	avg, _ := stats.Mean(data)
-	max, _ := stats.Max(data)
-	stddev, _ := stats.StandardDeviation(data)
-	return min, avg, max, stddev
+// Parse IP and TCP headers from ICMP Response Payload
+func getHeadersFromICMPResponsePayload(data []byte) (*layers.IPv4, *layers.TCP, error) {
+	if len(data) < 1 {
+		return nil, nil, errors.New("Invalid IP header")
+	}
+	ipHeaderLength := int((data[0] & 0x0F) * 4)
+
+	if len(data) < ipHeaderLength {
+		return nil, nil, errors.New("Length of received ICMP packet too short to decode IP")
+	}
+	ip := layers.IPv4{}
+	tcp := layers.TCP{}
+	ipErr := ip.DecodeFromBytes(data[0:], gopacket.NilDecodeFeedback)
+	tcpErr := tcp.DecodeFromBytes(data[ipHeaderLength:], gopacket.NilDecodeFeedback)
+
+	if ipErr != nil && tcpErr != nil {
+		return nil, nil, ipErr
+	}
+
+	return &ip, &tcp, nil
+}
+
+// Returns current time, used to reset counter timer
+func resetTimerForCounter() time.Time {
+	return time.Now()
+}
+
+// Check if tracerouteHopTimeout has been reached since the timeToCheck was set
+func hasTracerouteHopTimedout(timeToCheck time.Time) bool {
+	timenow := time.Now()
+	if timenow.Sub(timeToCheck) >= tracerouteHopTimeout {
+		return true
+	}
+	return false
+}
+
+// Check if a particular IP Id (uint16 in layers.IPv4) is in the slice of IP Id's we have sent with a particular TTL value
+func sliceContains(slice []SentPacketData, value uint16) bool {
+	for _, v := range slice {
+		if v.HopIPId == value {
+			return true
+		}
+	}
+	return false
+}
+
+func getSentTimestampfromIPId(sentDataSlice []SentPacketData, ipid uint16) (time.Time, error) {
+	for _, v := range sentDataSlice {
+		if v.HopIPId == ipid {
+			return v.HopSentTime, nil
+		}
+	}
+	return time.Now(), errors.New("IP Id not in sent packets")
 }
 
-// Function that sends out TcpPing and returns RTT
-func sendTcpPing(dst string, seq uint64, timeout time.Duration) float64 {
-	startTime := time.Now()
-	conn, err := net.DialTimeout("tcp", dst, timeout)
-	endTime := time.Now()
-	if err == nil || strings.Contains(err.Error(), "connection refused") {
-		if err == nil {
-			defer conn.Close()
+// Listen on the provided pcap handler for packets sent
+func recvPackets(uuid string, handle *pcap.Handle, serverIP string, clientIP string, hops chan HopRTT) {
+	var ipIdHop = make(map[int][]SentPacketData)
+	packetStream := gopacket.NewPacketSource(handle, handle.LinkType())
+	counter := beginTTLValue
+	timerPerHopPerUUID[uuid] = time.Now()
+	var hopRTTVal time.Duration
+	for packet := range packetStream.Packets() {
+		// if a particular hop is taking too long, then return nil and move on to sending packets to find the next hop
+		if hasTracerouteHopTimedout(timerPerHopPerUUID[uuid]) {
+			counter += 1
+			timerPerHopPerUUID[uuid] = resetTimerForCounter()
+			// Must produce and return a <nil> value otherwise empty structs containing net.IP cannot be compared
+			var empty net.IP
+			hops <- HopRTT{empty, 0}
 		}
```

Comment:
One comment/issue I have found here: This goroutine only moves on to the next hop (say, `n+1`) if there is no response from a particular hop `n` and there are other packets (to/from the source) seen on the pcap handle, i.e. from the packet stream we see some packet and we realize we have been waiting beyond the `tracerouteHopTimeout` and so, move on to next hop. But, if there are no such packets seen on the handle, the goroutine will keep waiting until it sees a packet to signal this timeout. 

Is there an external timer we can create and say, look for packets **for this hop** (`for packet := range packetStream.Packets()`) only until `tracerouteHopTimeout`. If timed out, do (reset timer, increment counter, send back "empty" `HopRTT` continue listening for packets) and then, start looking for packets for the next hop.

Reply:
Something like the following should work:
```go
ticker := time.NewTicker(tickerFrequency)
for {
  select {
    case pkt := <- packetStream.Packets():
      // Deal with packets.
    case <-ticker.C:
      // Check if there are timeouts.
  }
}
```
The ticker fires every, say, 500 ms and allows us to check if any packets timed out.

Reply:
After a bit of debugging and thinking about how to implement this logic in the recvPackets goroutine. I realized implementing it in the sending logic would make much more sense. I've done the following in [b0707ed]:(https://github.com/brave-experiments/proxy-detection-server/pull/5/commits/b0707eddffa390aa9c8571f416911f63dcf51dc1)

Send packet with TTL `t`, update a global map that keeps track of which TTL number we are currently on, then fire a ticker for `tracerouteHopTimeout`. Using the `select` goroutine wait, look for either data about the responding hop from `recvPackets` function or for the ticker to fire. If ticker fires, then enter empty hop in our result, and continue on to the next hop. 

I think this is much cleaner than trying to deal with it on the receiving side, as we will have to `.Reset(t)` the ticker each time we receive the ICMP error (that we are looking for) for the current hop.